cornell-dti/course-plan
# Self Check Frontend Finish

### Summary 

This pull request puts the finishing touches on the Frontend for the Self Check add modal and the self check courses on the requirements sidebar. ([Notion](https://www.notion.so/courseplan/82bf355c4b7b49f99691eb9f3afb203c?v=5da9d646f3ef424bbceb1ca0e8e00cf8&p=0c4db7e8431845c7bbda1343693dc232))

- [x] Filter out courses from self-check dropdowns if they are assigned to another requirement (if that req and this req are both not double countable)
- [x] Add a placeholder on the self check add modal for season/year instead of defaulting to the current season/year
- [x] Styling updates to match Figma (add req name in add modal title, increase margin around dropdowns, style toggleable dropdown the same as self check dropdown)
- [x] Resolve bugs (Hide dropdown if self check is fulfilled, show "Add Course #" for course self checks, replace unchanged isFulfilled with isCompleted which brings back some old reqs styling)

![image](https://user-images.githubusercontent.com/25535093/111922416-07659480-8a70-11eb-8193-992bb701f171.png)
![image](https://user-images.githubusercontent.com/25535093/111922567-f6695300-8a70-11eb-9469-bad41086f4e3.png)




Remaining TODOs:

- [ ] Resolve the bigger issue of courses assigned to self checks even if they fail the checkers (next Sprint)
- [ ] Resolve the two bugs on notion found in edit semester that now apply to the self check modal (next Sprint)

### Test Plan 

1. Ensure that a course assigned to a non-double countable requirement does not show up in any non-double countable self check dropdowns
2. Ensure that a course assigned to a double countable req shows up in other self-check dropdowns. A great test for this is dragging a course from the probability req for CS to your plan, and then opening a self-check dropdown.
3. Ensure a course assigned to a non-double countable req still shows up in a double-countable self check. **NOTE:** I have not tested this, do we have any examples where this is the case?
4. Ensure self-check dropdowns cannot be added to for self-checks that are fulfilled. **NOTE:** I checked this with courses, but I don't think we have a credit self-check right now? Is there a way to test this?
5. Ensure the self-check add modal matches designs, that the default semester/year is "Select," and that adding it adds it to the correct semester
6. Ensure the req bar better matches styles now, especially for self-check and toggleable dropdowns, in addition to margins.
7. Ensure that new semester and edit semester work like they do on the web app (since some common code was changed).

### Notes 

I was unsure of how best to handle not showing courses in the self-check dropdowns depending on double counting cases, so the code is heavily commented and might not be that clean. Please let me know if I should remove any of my comments or handle this in a better way!

Also just found a bug when making this PR that I will investigate, where this ECON course fulfills Probability and Technical but only shows up for one.
![image](https://user-images.githubusercontent.com/25535093/111922509-95da1600-8a70-11eb-9342-37a7acd22d85.png)
